hpy/devel/src/runtime/argparse.c

```diff
@@ -78,6 +78,11 @@
  *     `Truth Value Testing <https://docs.python.org/3/library/stdtypes.html#truth>`_
  *     for more information about how Python tests values for truth.
  *
+ * ``s (unicode) [const char*]``
+ *      Convert a Python unicode object to a C string. The lifetime of the C string
+ *      is tied to the underlying Python object (TODO: should be tied to the handle
```

Comment:
IMO, the lifetime of the pointer should be tied to the handle.
The alternative is that it's a fresh "copy" and the caller then owns the pointer and must dealloc itself.
@antocuni @hodgestar @rlamy  What do you think?

Reply:
@fangerer Agreed, with the caveat that I have no idea how hard those will be to implement.

Reply:
I would not _enforce_ it for the sake of performance on CPython. For GraalPython (and maybe also PyPy) it's easier to attach the pointer to the handle and also free it when the handle is closed.
For CPython, we can still always return the same pointer. It's just important that the user does not rely on the pointer being valid after the handle was closed.

The second suggestion (i.e. the caller gets a fresh pointer and owns it) is probably a bad idea because that means we need allocate a fresh string and copy contents every time (even in CPython).

Reply:
I updated the documentation to

> the pointer is valid only until the corresponding HPy handle is closed

I think that it's fine if in the non-debug mode the pointer stays valid. No one should rely on that and ideally the debug mode should catch any errors. I can take a look at enforcing that in debug mode as one of the next things. For now I think the specification is enough.

Reply:
Moreover, I think this could be turned into more generic rule that the lifetime of any pointers returned by HPy API is tied to the HPy handle that "owns" them. 

It will be more fun with mutable data if there will be any such API. There it is also a question of when any changes to the data should be visible to other handles referencing the same object.

Reply:
We already had a similar discussion a while ago when we were discussing the semantics of `HPy_AsStruct` (formerly `HPy_Cast`).
I was bringing up that if we would (eventually) have multi-threading then we would need to define this properly. In CPython, it's just shared memory.

Reply:
I've created https://github.com/hpyproject/hpy/issues/236 for the debug mode enhancement.